bundles/org.openhab.binding.bmwconnecteddrive/src/main/java/org/openhab/binding/bmwconnecteddrive/internal/handler/ConnectedDriveProxy.java

```diff
@@ -0,0 +1,314 @@
+import org.eclipse.jetty.util.UrlEncoded;
+import org.openhab.binding.bmwconnecteddrive.internal.ConnectedDriveConfiguration;
+import org.openhab.binding.bmwconnecteddrive.internal.VehicleConfiguration;
+import org.openhab.binding.bmwconnecteddrive.internal.dto.NetworkError;
+import org.openhab.binding.bmwconnecteddrive.internal.dto.auth.AuthResponse;
+import org.openhab.binding.bmwconnecteddrive.internal.handler.simulation.Injector;
+import org.openhab.binding.bmwconnecteddrive.internal.utils.BimmerConstants;
+import org.openhab.binding.bmwconnecteddrive.internal.utils.Constants;
+import org.openhab.binding.bmwconnecteddrive.internal.utils.Converter;
+import org.openhab.binding.bmwconnecteddrive.internal.utils.ImageProperties;
+import org.openhab.core.io.net.http.HttpClientFactory;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.gson.JsonSyntaxException;
+
+/**
+ * The {@link ConnectedDriveProxy} This class holds the important constants for the BMW Connected Drive Authorization.
+ * They
+ * are taken from the Bimmercode from github {@link https://github.com/bimmerconnected/bimmer_connected}
+ * File defining these constants
+ * {@link https://github.com/bimmerconnected/bimmer_connected/blob/master/bimmer_connected/account.py}
+ * https://customer.bmwgroup.com/one/app/oauth.js
+ *
+ * @author Bernd Weymann - Initial contribution
+ * @author Norbert Truchsess - edit & send of charge profile
+ */
+@NonNullByDefault
+public class ConnectedDriveProxy {
+    private final Logger logger = LoggerFactory.getLogger(ConnectedDriveProxy.class);
+    private final Token token = new Token();
+    private final HttpClient httpClient;
+    private final HttpClient authHttpClient;
+    private final String legacyAuthUri;
+    private final ConnectedDriveConfiguration configuration;
+
+    /**
+     * URLs taken from https://github.com/bimmerconnected/bimmer_connected/blob/master/bimmer_connected/const.py
+     */
+    final String baseUrl;
+    final String vehicleUrl;
+    final String legacyUrl;
+    final String vehicleStatusAPI = "/status";
+    final String lastTripAPI = "/statistics/lastTrip";
+    final String allTripsAPI = "/statistics/allTrips";
+    final String chargeAPI = "/chargingprofile";
+    final String destinationAPI = "/destinations";
+    final String imageAPI = "/image";
+    final String rangeMapAPI = "/rangemap";
+    final String serviceExecutionAPI = "/executeService";
+    final String serviceExecutionStateAPI = "/serviceExecutionStatus";
+
+    public ConnectedDriveProxy(HttpClientFactory httpClientFactory, ConnectedDriveConfiguration config) {
+        httpClient = httpClientFactory.getCommonHttpClient();
+        authHttpClient = httpClientFactory.createHttpClient(AUTH_HTTP_CLIENT_NAME);
+        authHttpClient.setFollowRedirects(false);
+        try {
+            authHttpClient.start();
+        } catch (Exception e) {
+            logger.warn("Auth Http Client cannot be started");
```

Comment:
Right. After some thoughts I don't think it's the right place. I moved this into *updateToken* so if client isn't started it will be tried again for recovery. If *updateToken* fails correpsonding *callback.onError* is called and thing will be set to *OFFLINE*.